astropy/cosmology/core.py

```diff
@@ -1895,6 +1921,175 @@ def _T_hypergeometric(self, x):
         from scipy.special import hyp2f1
         return 2 * np.sqrt(x) * hyp2f1(1./6, 1./2, 7./6, -x**3)
 
+    def _dS_age(self, z):
+        """ Age of the universe in Gyr at redshift ``z``.
+
+        The age of a de Sitter Universe is infinite.
+
+        Parameters
+        ----------
+        z : array-like
+          Input redshifts.
+
+        Returns
+        -------
+        t : `~astropy.units.Quantity`
+          The age of the universe in Gyr at each input redshift.
+
+        See Also
+        --------
+        z_at_value : Find the redshift corresponding to an age.
+        """
+        if isiterable(z):
+            z = np.asarray(z)
+
+        # Add the Hubble time * z to get the right units in the right shape
+        return np.inf + self._hubble_time * z
```

Comment:
Optional suggestion: To me that's an odd an unclear way to set the shape and units, and I could imagine that changing in the future and catching us by surprise.  Could we do it in a more straightforward fashion?  I might make a inf_like non-member, non exported function and use that, then add the units in there.

Reply:
I can see improving the documentation by making it a function, but I don't see an obvious way to actually do the work itself.  Would you prefer:
A)
```
        return inf_likeA(self._hubble_time * z)

    def inf_likeA(x):
        """Return the shape and units of x with value infinity."""
       # The + is necessary to avoid np.inf * 0 => np.nan
        return np.inf + self._hubble_time * z
```

B)
```
        return inf_likeB(z, self._hubble_time)

    def inf_likeB(x, unit):
        """Return the shape of x with 'unit' and value infinity."""
        return np.inf * np.ones_like(x) * unit
```

Reply:
Ah, or perhaps you were suggesting:
```
        return self._hubble_time * inf_likeC(z)

    def inf_likeC(x):
        """Return same shape as x with value of infinity."""
        return np.inf * np.ones_like(x)
```

Reply:
At least on my machine, the somewhat more cumbersome

def inf_like(x):
    e = np.empty_like(x)
    e[:] = np.inf
    return e

is the fastest.

Reply:
But that doesn't work for scalar x.

Reply:
But with the following modification it works:
```
def inf_like(x):
    e = np.empty_like(x)
    np.copyto(e, np.inf, casting='unsafe')
    return e
```
I looked up the code for `numpy.ones_like` to figure this out:
https://github.com/numpy/numpy/blob/v1.13.0/numpy/core/numeric.py#L253

Reply:
Although that code gives negative infinity for integer inputs.  This is a consequence of the  `casting='unsafe'` argument, which is used to avoid warnings about converting int to float.

Reply:
Yeah, good point.   Stupid numpy scalars.

Should have used np.full_like anyways, but that *also* doesn't work
properly for scalars.

I've never found anything much better than this sort of thing:

```
if np.isscalar(x):
  return np.inf
else:
  return np.full_like(x, np.inf)
```

bizarrely, np.full_like is slower than my original function, but it is much clearer.

Reply:
Lets stay away from arguments with names like 'unsafe'.

Reply:
Thanks, I implemented your last form for `inf_like` that uses `np.full_like` and the scalar check.